0005-kinesis-proxy-json.md

```diff
@@ -0,0 +1,306 @@
+the partition key, we ensure that all messages coming from a given TCP
+connection are routed to the same shard, ensuring that they are stored in the
+order that they are received.
+
+To ensure that the format of the JSON messages are documented, a JSON Schema
+document will be added to a new repository: mbta/schemas. By documenting them
+in a GitHub repo, we have several advantages over a purpose-built schema
+registry:
+
+- pull requests can be opened for new message types, providing documentation
+  and discussion as new messages are added
+- naming the schemas after the message types allows programmatic mapping of
+the message to the schema, for validation or other processing For example,
+the schema for the "com.mbta.ocs.raw-message" schema could live at
+https://github.com/mbta/schemas/blob/HEAD/com.mbta.ocs.raw-message.schema.json
+
+## Further Documentation
+
+- Kinesis high-level overview: https://docs.aws.amazon.com/streams/latest/dev/key-concepts.html
+- Kinesis API reference: https://docs.aws.amazon.com/kinesis/latest/APIReference/Welcome.html
+- Kinesis Enhanced Fan-out Consumer: https://docs.aws.amazon.com/streams/latest/dev/building-enhanced-consumers-api.html
+- `ex_aws_kinesis` (ExAws interface to the Kinesis API): https://docs.aws.amazon.com/kinesis/latest/APIReference/Welcome.html
+- Kinesis Client Library (KCL): https://docs.aws.amazon.com/streams/latest/dev/shared-throughput-kcl-consumers.html
+- Elixir Multi-Lang Daemon interface: https://github.com/AdRoll/exmld
+- `kcl_ex` (native Elixir KCL implementation): https://github.com/uberbrodt/kcl_ex
+- JSON Schema: https://jsonschema.org
+
+# Drawbacks
+
+Kinesis is a new technology that we're adding to our infrastructure. Even
+though it's mostly managed by Amazon, we'll need to develop some in-house
+expertise with building producers and consumers.
+
+Enhanced fan-out consumers are not well-supported by the existing
+Elixir/Erlang Kinesis adapters. The functionality does work (Paul Swartz was
+able to get it working during proof-of-concept testing) but we'll need to
+write some of our own code to support the functionality.
+
+Running Kinesis locally is easy enough with tools like LocalStack or
+Kinesialite, but more involved than running a TCP proxy. This might require
+some re-working of how the RTR integration system works when pushing logged
+data into a local instance for testing. Kinesialite doesn't support enhanced
+fan-out, so we'd need to support both models in our code.
+
+# Rationale and alternatives
+
+<!-- Why is this design the best in the space of possible designs?
+- What other designs have been considered and what is the rationale for not choosing them?
+- What is the impact of not doing this? -->
+
+AWS has tools to proxy messages from Kinesis into Kafka (and vice versa) so if
+we need to migrate in the future that is possible. If we had a lot more
+operational staff (or more experience with Kafka) this might point to a
```

Comment:
Do we have any more experience with Kinesis than we do with Kafka, at this point, or is it a question of Kafka being fundamentally more complicated to set up and use? The existence of MSK seems to negate the concern about needing more ops folks to run Kafka.